pkg/azureutils/azure_disk_utils.go

```diff
@@ -310,3 +328,95 @@ func checkDiskName(diskName string) bool {
 
 	return true
 }
+
+func GetMaxSharesAndMaxMountReplicaCount(parameters map[string]string) (int, int) {
+	maxShares := 1
+	maxMountReplicaCount := 0
+	for param, value := range parameters {
+		if strings.EqualFold(param, MaxSharesField) {
+			parsed, err := strconv.Atoi(value)
+			if err != nil {
+				klog.Warningf("failed to parse maxShares value (%s) to int, defaulting to 1: %v", value, err)
+			} else {
+				maxShares = parsed
+			}
+		} else if strings.EqualFold(param, MaxMountReplicaCountField) {
+			parsed, err := strconv.Atoi(value)
+			if err != nil {
+				klog.Warningf("failed to parse maxMountReplica value (%s) to int, defaulting to 0: %v", value, err)
+			} else {
+				maxMountReplicaCount = parsed
+			}
+		}
+	}
+
+	if maxShares-1 < maxMountReplicaCount {
+		klog.Warningf("maxMountReplicaCount cannot be larger than maxShares - 1: defaulting to maxShares - 1")
+		maxMountReplicaCount = maxShares - 1
+	}
+	return maxShares, maxMountReplicaCount
+}
+
+func GetAzVolumePhase(phase v1.PersistentVolumePhase) v1alpha1.AzVolumePhase {
+	return v1alpha1.AzVolumePhase(phase)
+}
+
+func GetAzVolume(ctx context.Context, client client.Client, azDiskClient azDiskClientSet.Interface, azVolumeName, namespace string, useCache bool) (*v1alpha1.AzVolume, error) {
+	var azVolume *v1alpha1.AzVolume
+	var err error
+	if useCache {
+		azVolume = &v1alpha1.AzVolume{}
+		err = client.Get(ctx, types.NamespacedName{Name: azVolumeName, Namespace: namespace}, azVolume)
+	} else {
+		azVolume, err = azDiskClient.DiskV1alpha1().AzVolumes(namespace).Get(ctx, azVolumeName, metav1.GetOptions{})
```

Comment:
That cache may be stale - is there a way to update the cache with the newly retrieved object? It seems a waste that the cache will need a separate `Get` call to update itself.

Reply:
I will investigate into it

> @sunpa93 Since the source of this PR is currently your personal fork, the e2e, sanity & integration Github actions will not have access to the secrets needed to run them. (This is by design to avoid disclosing the secrets to arbitrary code introduced by PR.) Could instead push the branch from your fork to a corresponding branch in this repo and then resubmit the PR so that the tests can run?

Ah that makes sense. I will resolve the comments, push the changes to mandalay repo and create a new PR 👍

Reply:
As we had initially discussed during the early stages of building our controller, there doesn't seem to be a clear way of updating the cache with client update calls. 

controller-runtime's cache reader internally uses client-go's cached indexer and doesn't expose any methods to manually update them (https://github.com/kubernetes-sigs/controller-runtime/blob/v0.9.0/pkg/cache/internal/cache_reader.go#L40). 

So unless we create our own implementation of the client, our best bet here would be to operate under the assumption that the cache reads can be stale from time to time but will eventually get up to date. Also, I was able to get around most of stale cache problems by reusing updated version of the objects used to make update calls for our CRIs because these objects are updated upon successful update requests (https://github.com/kubernetes-sigs/controller-runtime/blob/fbf50b04fe17b0f091430d22557245d2aedabfd4/pkg/client/typed_client.go#L55:L71)

Reply:
Ah, I had forgotten about that discussion.